### PR TITLE
GC: Read commit explicitly when it is missing from prescanned commits

### LINKED ISSUE #8261: BUG: GC: GetGarbageCollectionCommits sometimes fails on missing starting point commit
Spark error:
```
io.lakefs.clients.sdk.ApiException: Message: Internal Server Error
HTTP response code: 500
HTTP response body: {"message":"save garbage collection commits: find expired commits: commit not found: 
```

Digging into the code - the failure came from `active_commits.go` `GetGarbageCollectionCommits` line 58:
```
	for startingPointIterator.Next() {
		startingPoint := startingPointIterator.Value()
		retentionDays := int(rules.DefaultRetentionDays)
		commitNode, ok := commitsMap[startingPoint.CommitID]
		if !ok {
			return nil, fmt.Errorf("%w: %s", ErrCommitNotFound, startingPoint.CommitID)
		}
```
Seems like the commit was not found in the commitsMap.
The retryer in the Scala API client is masking an error that might indicate a potential bug in lakeFS as we see this flow being retried several times until success or until retry exhaustion.
This might be an indication for a bug in the `GCStartingPointIterator` or the logic which populates `commitsMap`

### PR DESCRIPTION
**Pull request by arielshaqed**
GC _lists_ all commits, but it can still miss a concurrent commit.  When that happens, read it directly from KV.

The issue goes into detail why this is a good idea.

Fixes #8261.

### CODE REVIEW

**Comment by arielshaqed** [comment]
Will pull if **EITHER** reviewer approves, unless you approve but ask me not to pull.  THANKS!

File: pkg/graveler/retention/active_commits.go [hunk 1]
```diff
@@ -15,52 +15,109 @@ type CommitNode struct {
 	MetaRangeID  graveler.MetaRangeID
 }
 
-func NewCommitNode(creationDate time.Time, mainParent graveler.CommitID, metaRangeID graveler.MetaRangeID) CommitNode {
+// CommitsMap is an immutable cache of commits.  Each commit can be set
+// once, and will be read if needed.  It is *not* thread-safe.
+type CommitsMap struct {
+	ctx          context.Context
+	Log          logging.Logger
+	NumMisses    int64
```
Comment (Jonathan-Rosenberg) [inline pkg/graveler/retention/active_commits.go:23]: What about logging this?

**Review by Jonathan-Rosenberg — APPROVED** [review]
Great find! THANKS!
LVGTM!
One suggestion to monitor the occurrences of the described race.

File: pkg/graveler/retention/active_commits.go (hunk 1 above)
Comment (arielshaqed) [reply pkg/graveler/retention/active_commits.go:23]: Done!

**Review by arielshaqed — COMMENTED** [review]
Thanks!

**Pull request merged**